### DISCUSSION TITLE
Proposed Change to Etsy API v3 | Allow / Store External System Ids

### DISCUSSION THREAD

**Discussion by showtroylove** [etsy/open-api#947] [Ideas]
**updateListingInventory**, **uploadListingImage**, **updateVariationImages** would be affected, but generally these principles should be applied throughout the API.  

When you add inventory to a listing you specify your own **property_values -> value_ids**. Upon saving or some other event Etsy disposes or hides the ids, you supplied initially, from retrieval.   These ids and similar user supplied identifiers aid external systems to, among other things, programmatically supply an appropriate variation image(s) to their inventory.  

Etsy also disposes of any identifying properties  / ids when adding images to a listing.  To be clear, the API should be mindful that external system need to supply their own identifier to all Listings and associated objects.   

Steps to reproduce:

- **createDraftListing**() or **updateInventory** an existing listing.

- You're required to provide **property_id**, **values_ids** and values for the **property_values** array.
- Upon successful update review the resulting **property_values** -> **value_ids**.
- Notice they have all been replaced with Etsy's own Ids.    

Listing Images is the same, you have no way to supply any identifier other than the **filename** and **rank** (both required).  

- When you retrieve the Listing images you'll notices all possible identifiers have been removed including the **filename** you supplied.
- Conceivably an external system could use the "**rank**" as a positional way to identify an image, but that's a workaround - if you will.

API's _should_ inherently support external id's for its consumers.  There is little effort, it seems to me, on the API team's part and it could make a big difference for the consumer.   Please consider offer ids for external systems throughout your API.  Gratitude in advance for your consideration.

**Comment by Gareth064** [comment]
Etsy API developers will likely never add something to help store external ID's. The closet thing you have to that is the SKU on the Product/Variation.

The reason why they probably won't, is because the API is just a façade in front the the core Etsy systems and has no influence on how these core systems operate. This was mentioned a few times by Etsy "API" developers on here in other issues/discussions. They just do their best to work with what the core systems provide. Doesn't mean they can't take ideas, we have provided here, and pass them onto the core teams, but I am yeah to see anything drastically impactful happen in that space.

To help you out, although not ideal, but sounds like the only course of action here is to store the Etsy generated ID in your system with a mapping to your internal ID. Then the next time you need you do an update, you can swap the Etsy ID into the request instead of you're own. This is just a stab in the dark suggestion as I am "yet" to work with listings and variations etc. That is the next phase of my app!

**Reply by showtroylove** [reply]
@Gareth064 I'm employing the phrase, "it doesn't hurt to ask."   I been on both sides of the isle namely, the API team and the consumer side (as we are now).  As part of the API team, we listened and consider carefully the request of our user community (absent obligation to formally do so).  

As Etsy matures into a well-oiled machine that is Amazon we'll see a more streamline API.   I think contributors like us help shape what that future "streamlined" API will look like.  

I'm curious about the app you've mentioned in other post in this forum.  Are you building something you plan to release commercially or are you building it for your own personal use?

**Reply by Gareth064** [reply]
I 100% agree about asking. I have just been on the more skeptical side these days when it comes to this API. Devs have been asked for, what seems like logical and for the most part simple, updates to be made and it hasn't happened.

I'm a Systems Integration Developer by trade and know how hard it can be to build an API on top of a rigid system you have no control over. You either try to get super creative at the API later or do your best to convince the other Devs to make changes to the core system.

Unless it's changed in the past few months, the Etsy "API" Dev team were just a couple of people working on the new API by themselves and are not part of the wider Etsy Dev Team. This is likely why changes and updates are taking so long.

We really need Etsy to invest in the API team more, and integrate them (pun intended) better with the other core teams.

**Reply by Gareth064** [reply]
Oh and to note also, this GitHub repo has become a community support site now replacing what was the Google groups) and isn't something the Etsy team actively uses anymore to "support" app devs.

If you want new features or actual support which requires Etsy Dev involvement then you need to email them.

Etsy will likely jump into chats every now and then, or give announcements here, but they told me (when I emailed them) that this is for the community to support each other now.

I argued that Etsy should stick around in the repo and help as they have been since starting v3 development, but they told me that that isn't the plan for now. And we should just email them with issues, which then get tracked by their internal support system.

**Reply by showtroylove** [reply]
I think this idea is brilliant.  Your development team may reduce a substantial number of _trivial_ request for support by having such a forum in place.  Thereby freeing up value time and resources for product development.  Brilliant!  If the nature of work I do for my clients wasn't so proprietary - I'd start a forum for each of them.
